ClientRuntimes/NodeJS/ms-rest-azure/lib/azureServiceClient.js

```diff
@@ -57,73 +57,76 @@ function AzureServiceClient(credentials, options) {
-    return callback(new Error(util.format('Unexpected polling status code from long running operation \'%s\'', 
-      resultOfInitialRequest.response.statusCode)));
+  if (!resultOfInitialRequest.response) {
+    return callback(new Error('Missing resultOfInitialRequest.response'));
+  }
+
+  if (this._checkInitialRequestResponseStatusCodeFailed(resultOfInitialRequest)) {
+    return callback(new Error(util.format('Unexpected polling status code from long running operation \'%s\' for method \'%s\'',
+      resultOfInitialRequest.response.statusCode,
+	  resultOfInitialRequest.request.method)));
   }
+  
   var pollingState = null;
   try {
     pollingState = new PollingState(resultOfInitialRequest, this.longRunningOperationRetryTimeout);
   } catch (error) {
     callback(error);
   }
-  
   var resourceUrl = resultOfInitialRequest.request.url;
   this._options = options;
-  
+
   async.whilst(
-    //while condition
-    function () {
-      var finished = [LroStates.Succeeded, LroStates.Failed, LroStates.Canceled].some(function (e) {
-        return pollingState.status === e;
+    function() {
+      var finished = [LroStates.Succeeded, LroStates.Failed, LroStates.Canceled].some(function(e) {
+        return e === pollingState.status;
       });
       return !finished;
     },
-    //while loop body
     function (callback) {
-      setTimeout(function () {
+      setTimeout(function() {
         if (pollingState.azureAsyncOperationHeaderLink) {
-          self._updateStateFromAzureAsyncOperationHeader(pollingState, false, function (err) {
+          self._updateStateFromAzureAsyncOperationHeader(pollingState, true, function(err) {
             return callback(err);
           });
         } else if (pollingState.locationHeaderLink) {
-          self._updateStateFromLocationHeaderOnPut(pollingState, function (err) {
+          self._updateStateFromLocationHeader(resultOfInitialRequest.request.method, pollingState, function(err) {
             return callback(err);
           });
-        } else {
-          self._updateStateFromGetResourceOperation(resourceUrl, pollingState, function (err) {
+        } else if (resultOfInitialRequest.request.method === 'PUT') {
+          self._updateStateFromGetResourceOperation(resourceUrl, pollingState, function(err) {
             return callback(err);
           });
+        } else {
+          return callback(new Error('Location header is missing from long running operation.'));
         }
       }, pollingState.getTimeout());
     },
     //when done
     function (err) {
       if (pollingState.status === LroStates.Succeeded) {
-        if (!pollingState.resource) {
-          self._updateStateFromGetResourceOperation(resourceUrl, pollingState, function (err) {
+        if ((pollingState.azureAsyncOperationHeaderLink || !pollingState.resource) && resultOfInitialRequest.request.method !== 'DELETE' && resultOfInitialRequest.request.method !== 'POST') {
```

Comment:
@csmengwan This would be easier to read if we did two things (1) assign resultOfInitialRequest.request.method to a local variable (called initialRequestMethod, and (2) make these checks positive (initialRequestMethod === 'PUT' || initialRequestMethod === 'PATCH'.  I would aslo suggest just making this request for a PUT or PATCH regardless of which polling method we use - the worst that can happen is an extra Get request.  Additionally, this seems a little unsafe when using the Location header for polling.